cirq/contrib/acquaintance/gates.py

```diff
@@ -0,0 +1,311 @@
+        acquaintance_size: An int indicating the locality of the logical gates
+            desired; used to keep track of this while nesting. If 0, no
+            acquaintance gates are inserted.
+
+    Attributes:
+        part_lens: See above
+        acquaintance_size: See above.
+        swap_gate: The gate used to swap logical indices.
+    """
+
+    def __init__(self,
+                 part_lens: Sequence[int],
+                 acquaintance_size: int=0,
+                 swap_gate: Gate=SWAP
+                 ) -> None:
+        if len(part_lens) < 2:
+            raise ValueError('part_lens must have length at least 2.')
+        self.part_lens = tuple(part_lens)
+        self.acquaintance_size = acquaintance_size
+        self.swap_gate = swap_gate
+
+
+    def default_decompose(self, qubits: Sequence[QubitId]) -> OP_TREE:
+        qubit_to_position = {q: i for i, q in enumerate(qubits)}
+        mapping = {q: i for i, q in enumerate(qubits)}
+        parts = []
+        q = 0
+        for part_len in self.part_lens:
+            parts.append(list(qubits[q: q + part_len]))
+            q += part_len
+        n_parts = len(parts)
+        op_sort_key = (lambda op:
+                qubit_to_position[min(op.qubits, key=qubit_to_position.get)] %
+                self.acquaintance_size)
+        layers = {'posterior_interstitial': []
+                  } # type: Dict[str, List[Operation]]
+        for layer_num in range(n_parts):
+            layers['prior_interstitial'] = layers['posterior_interstitial']
+            for l in ('pre', 'intra', 'post', 'posterior_interstitial'):
+                layers[l] = []
+            for i in range(layer_num % 2, n_parts - 1, 2):
+                left_part, right_part = parts[i:i+2]
+                acquaint_and_shift(parts=(left_part, right_part),
+                                   layers=layers,
+                                   acquaintance_size=self.acquaintance_size,
+                                   swap_gate=self.swap_gate,
+                                   mapping=mapping)
+
+                parts_qubits = list(left_part + right_part)
+                parts[i] = parts_qubits[:len(right_part)]
+                parts[i + 1] = parts_qubits[len(right_part):]
+            layers['prior_interstitial'].sort(key=op_sort_key)
+            for l in ('prior_interstitial', 'pre', 'intra', 'post'):
+                yield layers[l]
+        layers['posterior_interstitial'].sort(key=op_sort_key)
+        yield layers['posterior_interstitial']
+
+        # finish reversal
+        for part in reversed(parts):
+            part_len = len(part)
+            if part_len > 1:
+                positions = [mapping[q] for q in part]
+                offset = min(positions)
+                permutation = {
+                        i: (mapping[q] - offset) for i, q in enumerate(part)}
+                reverse_permutation = {
+                        i: part_len - p - 1 for i, p in permutation.items()}
```

Comment:
Modular arithmetic makes it very easy to keep values in range: `(-p - 1) % part_len`. Simple enough that you can merge it into the above line `(offset - mapping[q] - 1) % part_len`